### DISCUSSION TITLE
How to get Complete Kubernetes Version from RKE(Rancher)? How to detect RKE Cluster?

### DISCUSSION THREAD

**Discussion by KiranBodipi** [rancher/rke2#4953] [Q&A] [answered]
I have a RKE Cluster after I export the cluster.yaml via CLI and Query the cluster via CLI with kubectl version command, ending up with below information
```
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3", GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean", BuildDate:"2023-06-14T09:53:42Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"26", **GitVersion:"v1.26.8"**, GitCommit:"395f0a2fdc940aeb9ab88849e8fa4321decbf6e1", GitTreeState:"clean", BuildDate:"2023-08-24T00:43:07Z", GoVersion:"go1.20.7", Compiler:"gc", Platform:"linux/amd64"}
```

**Expected** : **GitVersion:"v1.26.8-rancher1-1"**
Actual: GitVersion:"v1.26.8"
In the same way when am querying the RKE2 Cluster, I see GitVersion:"v1.24.17", where as I was expecting full GitVersion as "v1.24.17+rke2r1"
Objective: In order to know which exact rancher kubernetes version(RKE/RKE2/K3s) is configured in cluster, am looking for this complete value? Is it a valid assumption in RKE/RKE2 cluster? Is there any other way or criteria to distinguish among these distributions?

**Comment by brandond** [comment]
Are you sure you're actually using RKE2 on the cluster in question? The output you provided does not look like what I would expect from this distribution. Here is what I see:
```console
brandond@dev01:~$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.3", GitCommit:"aef86a93758dc3cb2c658dd9657ab4ad4afc21cb", GitTreeState:"clean", BuildDate:"2022-07-13T14:30:46Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.17+rke2r1", GitCommit:"22a9682c8fe855c321be75c5faacde343f909b04", GitTreeState:"clean", BuildDate:"2023-08-24T04:20:55Z", GoVersion:"go1.20.7 X:boringcrypto", Compiler:"gc", Platform:"linux/amd64"}
```

Note: `GitVersion:"v1.24.17+rke2r1"` and  `GoVersion:"go1.20.7 X:boringcrypto"` do not match your output, which indicates that you are not querying an RKE2 server.

**Reply by KiranBodipi** [reply]
Thanks for the quick response @brandond 
I am not sure on the RKE2 cluster info as I don't have the cluster anymore. Is RKE2/RKE cluster always give the GitVersion "v1.24.17+rke2r1" with rancher distribution info ("rke2r1") on executing `kubectl version`? as I doubt my cluster configuration.

**Reply by brandond** [reply]
yes. rke2 will say +rke2r1, k3s will say +k3s1. Sometimes we do a second release of a patch, in which case it will end with 2 instead. If you don't see this you're not using rke2 or k3s.

**Reply by KiranBodipi** [reply]
While I was going through different release versions of RKE, I found that the RKE Kubernetes versions contain 'rancher1' or 'rancher2' as listed below:

v1.27.6-rancher1-1
v1.26.9-rancher1-1
v1.25.14-rancher1-1
v1.24.17-rancher1-1
v1.23.16-rancher2-3
Is this always the case, or can it contain 'rancher3', 'rancher4', or other versions (e.g., 'rancherX')? I am looking for criteria to detect the Rancher(RKE) platform, and your insights would be much appreciated.

**Reply by brandond** [reply]
Same answer, it depends on how many releases they do on that patch.

**Reply by brandond** [reply]
What exactly are you trying to do? Have you seen https://github.com/rancher/kubernetes-provider-detector ?

**Reply by KiranBodipi** [reply]
Well, am trying to identify a valid criteria to decide the Rancher distribution(RKE/RKE2/K3s) based on the Rancher Kubernetes Version.
As the clusters RKE/RKE2 provisioned by us doesn't have the complete GitVersion along with distribution info like rke2r1/rancher1-1/k3s1 in addition to major version, trying to know whether this is by design or there is an issue with cluster configuration.
Have you seen https://github.com/rancher/kubernetes-provider-detector ? Thanks for letting me know on this, based on node annotations also we can detect rancher distribution.

**Reply by brandond** [reply]
> As the clusters RKE/RKE2 provisioned by us doesn't have the complete GitVersion along with distribution info like rke2r1/rancher1-1/k3s1

It sure sounds like you're not running a Rancher Kubernetes distribution... How exactly are you launching these clusters?

**Reply by KiranBodipi** [reply]
Hey @brandond Well, am not much aware of how they are internally configuring these clusters. This is handled by some other team.But now, able to see the complete gitversion with distribution too.
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.9+rke2r1", GitCommit:"d1483fdf7a0578c83523bc1e2212a606a44fd71d", GitTreeState:"clean", BuildDate:"2023-09-13T20:34:35Z", GoVersion:"go1.20.8 X:boringcrypto", Compiler:"gc", Platform:"linux/amd64"}

But in case of RKE, was not able to see the complete gitversion with distribution.
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.8", GitCommit:"395f0a2fdc940aeb9ab88849e8fa4321decbf6e1", GitTreeState:"clean", BuildDate:"2023-08-24T00:43:07Z", GoVersion:"go1.20.7", Compiler:"gc", Platform:"linux/amd64"}

We were suspecting that RKE Cluster may not give the distribution info by design. Let me know, if you have any idea on rke cluster providing gitversion along with distribution info on kubectl version execution.

**Reply by KiranBodipi** [reply, marked as answer]
As per the logic in employed https://github.com/rancher/kubernetes-provider-detector/blob/master/providers/rke.go, it seems like there will not be any distribution info for the RKE in Server Version Info. We can rely on the following annotations of node ("rke.cattle.io/external-ip","rke.cattle.io/internal-ip") to decide RKE cluster.
Thanks for all the help!
